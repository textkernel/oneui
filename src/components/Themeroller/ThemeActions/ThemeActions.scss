.ThemeActions {
    display: flex;
    flex-wrap: wrap;
    padding: var(--spacing-5x) var(--spacing-2x) var(--spacing-2x) 0;

    &__button {
        margin-right: var(--spacing-3x);
        margin-bottom: var(--spacing-3x);
    }

    &__icon {
        padding-right: var(--spacing-normal);
    }

    &__closeIcon {
        padding-right: var(--spacing-normal);
        font-size: 16px;
    }
}
