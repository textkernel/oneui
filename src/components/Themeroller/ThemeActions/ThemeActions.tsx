import * as React from 'react';
import { FaDownload, FaUpload, FaUndo } from 'react-icons/fa';
import { MdClose } from 'react-icons/md';
import { bem } from '../../../utils';
import { Button, FileButton } from '../../Buttons';
import { Tooltip } from '../../Tooltip';
import styles from './ThemeActions.scss';

interface Props {
    /** label for reset button */
    resetLabel?: string;
    /** label for active reset button */
    resetActiveLabel?: string;
    /** label for download button */
    downloadLabel?: string;
    /** label for tooltip download button */
    downloadTooltipLabel?: string;
    /** label for choose file button */
    fileLabel?: string;
    /** if the download button should be disabled */
    downloadDisabled?: boolean;
    /** callback is called when reset was called */
    onReset?: () => void;
    /** callback is called when active reset was called */
    onActiveReset?: () => void;
    /** callback is called when download was called */
    onDownload?: () => void;
    /** callback is called when choose file was called */
    onFileChange?: (e: React.ChangeEvent<HTMLInputElement>, content: string) => void;
}

const { block, elem } = bem('ThemeActions', styles);

export const ThemeActions: React.FC<Props> = ({
    resetLabel,
    resetActiveLabel,
    downloadLabel,
    fileLabel,
    downloadTooltipLabel,
    downloadDisabled,
    onReset,
    onActiveReset,
    onDownload,
    onFileChange,
}) => {
    const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        if (onFileChange && e.target.files && e?.target?.files?.[0]) {
            const file = e.target.files[0];
            const reader = new FileReader();
            reader.onload = (r) => {
                if (r?.target?.result) {
                    if (typeof r.target.result === 'string') {
                        onFileChange(e, r.target.result);
                    }
                }
            };
            reader.readAsText(file);
        }
    };
    return (
        <div {...block()}>
            {resetLabel && onReset && (
                <Button
                    {...elem('button')}
                    type="reset"
                    size="small"
                    context="bad"
                    onClick={onReset}
                >
                    <>
                        <MdClose {...elem('closeIcon')} />
                        {resetLabel}
                    </>
                </Button>
            )}
            {resetActiveLabel && onActiveReset && (
                <Button
                    {...elem('button')}
                    type="reset"
                    size="small"
                    context="warning"
                    onClick={onActiveReset}
                >
                    <>
                        <FaUndo {...elem('icon')} />
                        {resetActiveLabel}
                    </>
                </Button>
            )}
            {fileLabel && onFileChange && (
                <div {...elem('button')}>
                    <FileButton
                        size="small"
                        context="good"
                        accept="application/JSON"
                        onChange={handleFileChange}
                    >
                        <>
                            <FaUpload {...elem('icon')} />
                            {fileLabel}
                        </>
                    </FileButton>
                </div>
            )}
            {downloadLabel && onDownload && (
                <Tooltip
                    content={downloadTooltipLabel}
                    placement="bottom"
                    disabled={!downloadDisabled}
                >
                    <span>
                        {/**
                         * Tooltip doesn't work when button is disabled,
                         * thus we need to wrapp button in any tag.
                         */}
                        <Button
                            {...elem('button')}
                            disabled={downloadDisabled}
                            size="small"
                            onClick={onDownload}
                        >
                            <>
                                <FaDownload {...elem('icon')} />
                                {downloadLabel}
                            </>
                        </Button>
                    </span>
                </Tooltip>
            )}
        </div>
    );
};

ThemeActions.displayName = 'ThemeActions';

ThemeActions.defaultProps = {
    resetLabel: '',
    downloadLabel: '',
    fileLabel: '',
    downloadDisabled: false,
    onReset: undefined,
    onDownload: undefined,
    onFileChange: undefined,
};
