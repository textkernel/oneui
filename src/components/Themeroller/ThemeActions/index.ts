export { ThemeActions, Props as ThemeActionsProps } from './ThemeActions';
