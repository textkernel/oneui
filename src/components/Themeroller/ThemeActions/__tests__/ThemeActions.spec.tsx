import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { ThemeActions } from '../ThemeActions';

describe('ThemeActions component', () => {
    let view: RenderResult;

    const getButtonByName = (inputName) => screen.getByRole('button', { name: `${inputName}` });

    it('should render component correctly without any actions', () => {
        view = render(<ThemeActions />);

        expect(view.container).toMatchSnapshot();
        expect(screen.queryAllByRole('button')).toHaveLength(0);
    });

    it('should render component correctly with all actions', () => {
        const onResetMock = jest.fn();
        const onDownloadMock = jest.fn();
        const onFileChangeMock = jest.fn();
        view = render(
            <ThemeActions
                downloadDisabled
                resetLabel="Reset"
                fileLabel="Choose file"
                downloadLabel="Download"
                onReset={onResetMock}
                onDownload={onDownloadMock}
                onFileChange={onFileChangeMock}
            />
        );

        expect(view.container).toMatchSnapshot();
        expect(getButtonByName('Reset')).toBeVisible();
        expect(getButtonByName('Download')).toBeVisible();
    });

    it('should invoke onReset callback when reset button is clicked', async () => {
        const onResetMock = jest.fn();
        const user = userEvent.setup();
        view = render(<ThemeActions resetLabel="Reset" onReset={onResetMock} />);

        await user.click(getButtonByName('Reset'));

        expect(onResetMock).toHaveBeenCalledTimes(1);
    });

    it('should invoke onDownload callback when download button is clicked', async () => {
        const onDownloadMock = jest.fn();
        const user = userEvent.setup();
        view = render(<ThemeActions downloadLabel="Download" onDownload={onDownloadMock} />);

        await user.click(getButtonByName('Download'));

        expect(onDownloadMock).toHaveBeenCalledTimes(1);
    });
});
