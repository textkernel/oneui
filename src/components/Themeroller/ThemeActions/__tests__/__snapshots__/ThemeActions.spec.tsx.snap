// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeActions component should render component correctly with all actions 1`] = `
<div>
  <div
    class="ThemeActions"
  >
    <button
      class="Button Button--size_small ThemeActions__button"
      context="danger"
      type="reset"
    >
      <svg
        class="ThemeActions__closeIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      Reset
    </button>
    <div
      class="ThemeActions__button"
    >
      <label
        class="Button Button--size_small"
      >
        <input
          accept="application/JSON"
          hidden=""
          type="file"
        />
        <svg
          class="ThemeActions__icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M296 384h-80c-13.3 0-24-10.7-24-24V192h-87.7c-17.8 0-26.7-21.5-14.1-34.1L242.3 5.7c7.5-7.5 19.8-7.5 27.3 0l152.2 152.2c12.6 12.6 3.7 34.1-14.1 34.1H320v168c0 13.3-10.7 24-24 24zm216-8v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h136v8c0 30.9 25.1 56 56 56h80c30.9 0 56-25.1 56-56v-8h136c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
          />
        </svg>
        Choose file
      </label>
    </div>
    <span>
      <button
        class="Button Button--size_small ThemeActions__button"
        disabled=""
        type="button"
      >
        <svg
          class="ThemeActions__icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M216 0h80c13.3 0 24 10.7 24 24v168h87.7c17.8 0 26.7 21.5 14.1 34.1L269.7 378.3c-7.5 7.5-19.8 7.5-27.3 0L90.1 226.1c-12.6-12.6-3.7-34.1 14.1-34.1H192V24c0-13.3 10.7-24 24-24zm296 376v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h146.7l49 49c20.1 20.1 52.5 20.1 72.6 0l49-49H488c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
          />
        </svg>
        Download
      </button>
    </span>
  </div>
</div>
`;

exports[`ThemeActions component should render component correctly without any actions 1`] = `
<div>
  <div
    class="ThemeActions"
  >
    
    
    
  </div>
</div>
`;
